src/linear_algebra/charpoly/basic.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import linear_algebra.free_module.finite
+import linear_algebra.matrix.charpoly.coeff
+
+/-!
+
+# Characteristic polynomial
+
+We define the characteristic polynomial of `f : M →ₗ[R] M`, where `M` is a finite and
+free `R`-module. The proof that `f.charpoly` is the characteristic polynomial of the matrix of `f`
+in any basis is in `linear_algebra/charpoly/to_matrix`.
+
+## Main definition
+
+* `linear_map.charpoly f` : the characteristic polynomial of `f : M →ₗ[R] M`.
+
+-/
+
+universes u v w
+
+variables {R : Type u} {M : Type v} [comm_ring R] [nontrivial R]
+variables [add_comm_group M] [module R M] [module.free R M] [module.finite R M] (f : M →ₗ[R] M)
+
+open_locale classical matrix
+
+noncomputable theory
+
+open module.free polynomial matrix
+
+namespace linear_map
+
+section basic
+
+/-- The characteristic polynomial of `f : M →ₗ[R] M`. -/
+def charpoly : polynomial R :=
+(to_matrix (choose_basis R M) (choose_basis R M) f).charpoly
+
+lemma charpoly_def :
+  f.charpoly = (to_matrix (choose_basis R M) (choose_basis R M) f).charpoly := rfl
+
+end basic
+
+section coeff
+
+lemma charpoly_monic : f.charpoly.monic := charpoly_monic _
+
+end coeff
+
+section cayley_hamilton
+
+/-- The Cayley-Hamilton Theorem, that the characteristic polynomial of a linear map, applied to
+the linear map itself, is zero. -/
+lemma aeval_self_charpoly : aeval f f.charpoly = 0 :=
+begin
+  apply (linear_equiv.map_eq_zero_iff (alg_equiv_matrix _).to_linear_equiv).1,
+  rw [alg_equiv.to_linear_equiv_apply, ← alg_equiv.coe_alg_hom,
+    ← polynomial.aeval_alg_hom_apply _ _ _, charpoly_def],
+  exact aeval_self_charpoly _,
+end
+
+lemma is_integral : is_integral R f := ⟨f.charpoly, ⟨charpoly_monic f, aeval_self_charpoly f⟩⟩
+
+lemma minpoly_dvd_charpoly {K : Type u} {M : Type v} [field K] [add_comm_group M] [module K M]
+  [finite_dimensional K M] (f : M →ₗ[K] M) : minpoly K f ∣ f.charpoly :=
+minpoly.dvd _ _ (aeval_self_charpoly f)
+
+variable {f}
+
+lemma charpoly_coeff_zero_of_injective (hf : function.injective f) : (minpoly R f).coeff 0 ≠ 0 :=
```

Comment:
the lemma name mentions `charpoly`, but the statement is instead on `minpoly`.

Reply:
Thanks!